hudi-common/src/main/java/org/apache/hudi/common/table/view/RocksDbBasedFileSystemView.java

```diff
@@ -269,41 +269,51 @@ protected Option<Pair<String, CompactionOperation>> getPendingLogCompactionOpera
 
   @Override
   protected boolean isPartitionAvailableInStore(String partitionPath) {
-    String lookupKey = schemaHelper.getKeyForPartitionLookup(partitionPath);
-    Serializable obj = rocksDB.get(schemaHelper.getColFamilyForStoredPartitions(), lookupKey);
-    return obj != null;
+    try {
+      readLock.lock();
+      String lookupKey = schemaHelper.getKeyForPartitionLookup(partitionPath);
+      Serializable obj = rocksDB.get(schemaHelper.getColFamilyForStoredPartitions(), lookupKey);
+      return obj != null;
+    } finally {
+      readLock.unlock();
+    }
   }
 
   @Override
   protected void storePartitionView(String partitionPath, List<HoodieFileGroup> fileGroups) {
-    LOG.info("Resetting and adding new partition ({}) to ROCKSDB based file-system view at {}, Total file-groups={}",
-        partitionPath, config.getRocksdbBasePath(), fileGroups.size());
-
-    String lookupKey = schemaHelper.getKeyForPartitionLookup(partitionPath);
-    rocksDB.delete(schemaHelper.getColFamilyForStoredPartitions(), lookupKey);
-
-    // First delete partition views
-    rocksDB.prefixDelete(schemaHelper.getColFamilyForView(),
-        schemaHelper.getPrefixForSliceViewByPartition(partitionPath));
-    rocksDB.prefixDelete(schemaHelper.getColFamilyForView(),
-        schemaHelper.getPrefixForDataFileViewByPartition(partitionPath));
-
-    // Now add them
-    fileGroups.forEach(fg ->
-        rocksDB.writeBatch(batch ->
-            fg.getAllFileSlicesIncludingInflight().forEach(fs -> {
-              rocksDB.putInBatch(batch, schemaHelper.getColFamilyForView(), schemaHelper.getKeyForSliceView(fg, fs), fs);
-              fs.getBaseFile().ifPresent(df ->
-                  rocksDB.putInBatch(batch, schemaHelper.getColFamilyForView(), schemaHelper.getKeyForDataFileView(fg, fs), df)
-              );
-            })
-        )
-    );
+    try {
+      writeLock.lock();
```

Comment:
the change is not necessary, the invoker already holds the read lock.

Reply:
This is actually the core of the problem. A read lock is fundamentally different than a write lock. The write lock is taken when we update the resource the lock is protecting.

Reply:
the real problem is the code snippet guarded by the read lock sometimes updates the state of the fs view, while in most of the cases, it is still correct if the underneath state is thread-safe or update idempotent.

Reply:
Right, but isn't that why the write-lock is generated? It can be used to solve this problem in a general way instead of forcing each map implementation to be thread-safe. Previously, I was suggesting to make the map impls thread safe but it was suggested to solve it at the view layer instead.

Reply:
> Previously, I was suggesting to make the map impls thread safe but it was suggested to solve it at the view layer instead.

Did you have benchmark numbers for the synchronized disk map?

Not sure if it makes sense if we just put the write lock in the direct modification method such as `#ensurePartitionsLoadedCorrectly`, and remove the read lock from it's invoker. The read lock is only for read APIs that does not do any modification to the fs state.

Reply:
The implementation I had was not correct. The correct implementation will require the use of read/write locks similar to this but that will force all instances of the disk map to use these locks, even during compaction where we don't have the multi-threaded case.

The read lock is required when reading but we can move the read lock even closer to where the store is accessed to minimize repeated code and reduce risk of new methods missing these checks.

Reply:
In the latest commit, I moved the write lock to `addFilesToView` and removed the code from these specific implementations.